### PR TITLE
Add Convolutional Dictionary Operator

### PR DESCRIPTION
**Pull request by koflera**
This adds a convolutional dictionary operator as a MRpro LinearOperator.

The operator can be defined as an "analysis" or a "synthesis" dictionary, and implements the following:

Synthesis:
$Ds:=\sum_{k=1}^K d_k \ast s_k$, where $s=[s_1, \ldots, s_K]^T$

Analysis:
$Hx:=[h_1\ast x, \ldots, h_k \ast x]^T$

Thereby, for inputs $s_k$ and a filter $d_k$, each of the convs in the operator can be used for the following different cases:

1.  $s_k \in\mathbb{R}^k,  s_k \in\mathbb{R}^k$
2.  $s_k \in\mathbb{R}^k,  s_k \in\mathbb{C}^k$: in this, case $d_k \ast s_k:= d_k \ast \mathrm{Re} s_k + \mathrm{i} d_k \ast \mathrm{Im} s_k$
3.  $d_k \in\mathbb{C}^k,  s_k \in\mathbb{C}^k$

This operator is essentially a wrapper that allows to use shapes of arbitrary inputs, where only the the last ndim (defined by the kernel size) are used for the convolution.

Only kernels for which all convolutional dimensions are odd are supported to avoid any disambiguities between the pixel location for the composite application of adjoint and forward. At least to the best of my knowledge, I have never seen use cases with kernels of 8x8 or so.

Also, currently only 1D, 2D and 3D convs are supported, as this essentially just wraps the `torch.nn.functional.conv` operators, but at some point, we could integrate the nd conv from #753.

### CODE REVIEW

**Comment by fzimmermann89** [comment]
@koflera is this the same (untested) ?

```python
from typing import Literal

import torch
from torchnd import adjoint_pad_nd, conv_nd, pad_nd

from mrpro.operators.LinearOperator import LinearOperator


class ConvAnalysisOp(LinearOperator):
    """Convolutional dictionary analysis operator."""

    def __init__(
        self,
        kernel: torch.Tensor,
        pad_mode: Literal["constant", "reflect", "replicate", "circular"] = "circular",
    ) -> None:
        super().__init__()
        self.kernel = kernel
        self.pad_mode = pad_mode

    def forward(self, x: torch.Tensor) -> tuple[torch.Tensor]:
        n_dim = self.kernel.ndim - 1
        pad = tuple(p for k in self.kernel.shape[1:] for p in (k // 2, k // 2))

        batch_shape = x.shape[:-n_dim]
        x = x.flatten(0, -n_dim - 1).unsqueeze(1)
        x = pad_nd(x, pad=pad, dims=tuple(range(-n_dim, 0)), mode=self.pad_mode)
        y = conv_nd(x, self.kernel.unsqueeze(1), dim=dims)
        y = y.unflatten(0, batch_shape).moveaxis(-n_dim - 1, 0)

        return (y,)

    def adjoint(self, x: torch.Tensor) -> tuple[torch.Tensor]:
        n_dim = self.kernel.ndim - 1
        pad = tuple(p for k in self.kernel.shape[1:] for p in (k // 2, k // 2))

        batch_shape = x.shape[1:-n_dim]
        x = x.moveaxis(0, -n_dim - 1).flatten(0, -n_dim - 2)
        y = conv_nd(x, self.kernel.unsqueeze(1).conj(), dim=tuple(range(-n_dim, 0)), transposed=True)
        y = adjoint_pad_nd(y, pad=pad, dims=dims, mode=self.pad_mode).squeeze(1)
        
        if batch_shape
          y = y.unflatten(0, batch_shape)

        return (y,)


class ConvSynthesisOp(LinearOperator):
    """Convolutional dictionary synthesis operator."""

    def __init__(
        self,
        kernel: torch.Tensor,
        pad_mode: Literal["constant", "reflect", "replicate", "circular"] = "circular",
    ) -> None:
        super().__init__()
        self.kernel = torch.nn.Parameter(
            kernel, requires_grad=kernel.requires_grad
        )
        self.pad_mode = pad_mode

    def forward(self, x: torch.Tensor) -> tuple[torch.Tensor]:
        n_dim = self.kernel.ndim - 1
        pad = tuple(p for k in self.kernel.shape[1:] for p in (k // 2, k // 2))

        batch_shape = x.shape[1:-n_dim]
        x = x.moveaxis(0, -n_dim - 1).flatten(0, -n_dim - 2)
        x = pad_nd(x, pad=pad, dims=dims, mode=self.pad_mode)
        y = conv_nd(x, self.kernel.unsqueeze(0), dim=tuple(range(-n_dim, 0)))
        y = y.squeeze(1)
        if batch_shape:
          y = y.unflatten(0, batch_shape)

        return (y,)

    def adjoint(self, x: torch.Tensor) -> tuple[torch.Tensor]:
        n_dim = self.kernel.ndim - 1
        pad = tuple(p for k in self.kernel.shape[1:] for p in (k // 2, k // 2))

        batch_shape = x.shape[:-n_dim]
        x = x.flatten(0, -n_dim - 1).unsqueeze(1)
        y = conv_nd(x, self.kernel.unsqueeze(0).conj(), dim=dims, transposed=True)
        y = adjoint_pad_nd(y, pad=pad, dims=tuple(range(-n_dim, 0)), mode=self.pad_mode)
        y = y.unflatten(0, batch_shape).moveaxis(-n_dim - 1, 0)

        return (y,)
```

**Comment by koflera** [comment]
> @koflera is this the same (untested) ?
> 
> ```python
> from typing import Literal
> 
> import torch
> from torchnd import adjoint_pad_nd, conv_nd, pad_nd
> 
> from mrpro.operators.LinearOperator import LinearOperator
> 
> 
> class ConvAnalysisOp(LinearOperator):
>     """Convolutional dictionary analysis operator."""
> 
>     def __init__(
>         self,
>         kernel: torch.Tensor,
>         pad_mode: Literal["constant", "reflect", "replicate", "circular"] = "circular",
>     ) -> None:
>         super().__init__()
>         self.kernel = kernel
>         self.pad_mode = pad_mode
> 
>     def forward(self, x: torch.Tensor) -> tuple[torch.Tensor]:
>         n_dim = self.kernel.ndim - 1
>         pad = tuple(p for k in self.kernel.shape[1:] for p in (k // 2, k // 2))
> 
>         batch_shape = x.shape[:-n_dim]
>         x = x.flatten(0, -n_dim - 1).unsqueeze(1)
>         x = pad_nd(x, pad=pad, dims=tuple(range(-n_dim, 0)), mode=self.pad_mode)
>         y = conv_nd(x, self.kernel.unsqueeze(1), dim=dims)
>         y = y.unflatten(0, batch_shape).moveaxis(-n_dim - 1, 0)
> 
>         return (y,)
> 
>     def adjoint(self, x: torch.Tensor) -> tuple[torch.Tensor]:
>         n_dim = self.kernel.ndim - 1
>         pad = tuple(p for k in self.kernel.shape[1:] for p in (k // 2, k // 2))
> 
>         batch_shape = x.shape[1:-n_dim]
>         x = x.moveaxis(0, -n_dim - 1).flatten(0, -n_dim - 2)
>         y = conv_nd(x, self.kernel.unsqueeze(1).conj(), dim=tuple(range(-n_dim, 0)), transposed=True)
>         y = adjoint_pad_nd(y, pad=pad, dims=dims, mode=self.pad_mode).squeeze(1)
>         
>         if batch_shape
>           y = y.unflatten(0, batch_shape)
> 
>         return (y,)
> 
> 
> class ConvSynthesisOp(LinearOperator):
>     """Convolutional dictionary synthesis operator."""
> 
>     def __init__(
>         self,
>         kernel: torch.Tensor,
>         pad_mode: Literal["constant", "reflect", "replicate", "circular"] = "circular",
>     ) -> None:
>         super().__init__()
>         self.kernel = torch.nn.Parameter(
>             kernel, requires_grad=kernel.requires_grad
>         )
>         self.pad_mode = pad_mode
> 
>     def forward(self, x: torch.Tensor) -> tuple[torch.Tensor]:
>         n_dim = self.kernel.ndim - 1
>         pad = tuple(p for k in self.kernel.shape[1:] for p in (k // 2, k // 2))
> 
>         batch_shape = x.shape[1:-n_dim]
>         x = x.moveaxis(0, -n_dim - 1).flatten(0, -n_dim - 2)
>         x = pad_nd(x, pad=pad, dims=dims, mode=self.pad_mode)
>         y = conv_nd(x, self.kernel.unsqueeze(0), dim=tuple(range(-n_dim, 0)))
>         y = y.squeeze(1)
>         if batch_shape:
>           y = y.unflatten(0, batch_shape)
> 
>         return (y,)
> 
>     def adjoint(self, x: torch.Tensor) -> tuple[torch.Tensor]:
>         n_dim = self.kernel.ndim - 1
>         pad = tuple(p for k in self.kernel.shape[1:] for p in (k // 2, k // 2))
> 
>         batch_shape = x.shape[:-n_dim]
>         x = x.flatten(0, -n_dim - 1).unsqueeze(1)
>         y = conv_nd(x, self.kernel.unsqueeze(0).conj(), dim=dims, transposed=True)
>         y = adjoint_pad_nd(y, pad=pad, dims=tuple(range(-n_dim, 0)), mode=self.pad_mode)
>         y = y.unflatten(0, batch_shape).moveaxis(-n_dim - 1, 0)
> 
>         return (y,)
> ```

yes, looks good. The only thing is that I would stick to excluding the case for complex-valued kernels and real-valued inputs for the following reasons.
1. i don't see any point in allowing that from a signal-processing perspective. Imagine if you want to do the approximation of a complex-valued image  x=Ds, but you only use real-valued sparse feature maps s?
2. It introduces ambiguities in the mathematical spaces. If the adjoint property should hold, then, for $D: V\rightarrow W$ we need to have $\langle Du, v\rangle_W = \langle u, D^H v\rangle_V$, but if the kernel is complex-valued, even for real-valued inputs, we have $W=\mathbb{C}^N$. But in general, $D^Hv \in \mathbb{C}^{NK}$, meaning that, in order to define the inner product in $V$, you need to use $V=\mathbb{C}^{NK}$ anyway by using tensors with zero imaginary components, so you are actually back in the complex-valued input and complex-valued kernel case.
3. even pytorch excludes the mixed cases. However, keeping the real-valued kernel and complex-valued inputs makes sense here, as we have a specific application in mind and a consistent way to handle it.

**Comment by koflera** [comment]
the remaining errors should all be related to mypy failures related to apparent typos, e.g.
`error: nd should be and` when using `pad_nd`

What is the preferred/recommended way to ignore these?

**Comment by fzimmermann89** [comment]
> the remaining errors should all be related to mypy failures related to apparent typos, e.g. `error: nd should be and` when using `pad_nd`
> 
> What is the preferred/recommended way to ignore these?

https://github.com/PTB-MR/mrpro/blob/fd1a24564a9261c0de94f44ba05bd43eb3242009/pyproject.toml#L227-L233

**Comment by fzimmermann89** [comment]
should work now (the pre-commit.ci error should fix itsel at some point i think..)

**Comment by fzimmermann89** [comment]
i added test for "no batch dim" and did a tiny bit of cleanup of the no-batch/batch logic.

**Comment by fzimmermann89** [comment]
There is at least one usecase (if i understand the equations correct) for complex-valued kernels and real valued inputs:
The equations are symmetric in kernel/input. So if you want to self supervised learn kernels, you can set up the Operator once with kernel=image, and apply with input=image.

I think there should not be any mroe ambiguities by also allowin the other mixed case as well. 
$\sum_{k=1}^K d_k \ast s_k$ and $[h_1\ast x, \ldots, h_k \ast x]^T$ only need  $\ast$ to be defined for mixed inputs.

I would just show a warning saying that the input the kernel is complex and thus the input is also interpreted as $\in C$ instead of an exception. But not a point I would fight over.

We should mention that this actually does a Correlation, not a convolution (not commutative) or flip the kernel on  creaction.


I made the kernel a tensor, not nn.Parameter to be able to use the output if a network there.

**Review by fzimmermann89 — APPROVED** [review]
Please have a last look if my changes broke anything and mention correlation in docstrings

**Review by fzimmermann89 — CHANGES_REQUESTED** [review]
something broke in merging?
some of my changes got overwritten in https://github.com/PTB-MR/mrpro/pull/981/changes/b12ba3e07d80e542c5b91a192430088a5e3cb7f5#diff-c3cc0af1f8c1fc1f2b6b7d4f4b0a664014d2b25859812b0b818f4f421d6a6abe

**Review by fzimmermann89 — APPROVED** [review]
lgtm
